src/static/js/chapter.js

```diff
@@ -256,6 +256,102 @@ function setDiscussionCount() {
+  if('IntersectionObserver' in window){
+    var indexScroller = document.querySelector('.index-scroller');
+
+    // Check if user has set reduced motion and only continue if not
+    var hasOSReducedMotion = window.matchMedia('(prefers-reduced-motion: reduce)').matches;
+    if (hasOSReducedMotion) {
+      console.log('User has set prefers-reduced-motion to ' + hasOSReducedMotion + ' so not highlighting the current section in chapter index');
+    } else {
+
+      // Check if 'position:sticky' is supported (as this is not great UX when not so don't bother)
+      // Add the sticky class (which sets 'position:sticky') and then test if that stuck :-)
+      indexScroller && indexScroller.parentNode && indexScroller.parentNode.classList.add('sticky');
+      var indexScrollerStyles = getComputedStyle(indexScroller.parentNode);
+      if (indexScrollerStyles && indexScrollerStyles.position) {
+        //Handle vendor prefixes (still required for Safari 12)
+        if (indexScrollerStyles.position.endsWith('sticky')) {
+
+          // Restrict the page height of the index to the page-height, as we're going to scroll this.
+          indexScroller.classList.add('page-height');
+
+          var indexItems = {};
+          var currentItem;
+
+          // Create a function to handle highlighting a new index item
+          // that will be called y bthe IntersectionObserver
+          function highlightIndexEntry(link) {
+            
+            var oldItem = currentItem;
+            var newItem = indexItems[link];
+            if (newItem){
+              newItem.classList.add('active');
+              if (oldItem && oldItem !== newItem) {
+                oldItem.classList.remove('active');
+              }
+              currentItem = newItem;
+
+              // If the index is too large to display in full then might need to change scroll
+              if (indexScroller.scrollHeight > indexScroller.clientHeight) {
+                var currentPosition = currentItem.scrollTop;
+                var currentNode = currentItem;
+                // Walk the node back up to the index-scroller to get the total position of the element
+                while (currentNode && !currentNode.parentNode.classList.contains('index-scroller')) {
+                  currentPosition = currentPosition + currentNode.offsetTop;
+                  currentNode = currentNode.parentNode;
+                }
+
+                // Show the current image in the middle of the screen
+                indexScroller.scrollTop = currentPosition - (indexScroller.clientHeight / 2);
+                  
+              }
+            }
+          }
+
+          // Set up a new Interstection Observer for when the title is 80% from the bottom of the page
+          var options = {
+            root: null,
+            rootMargin: "0px 0px -80% 0px",
+            threshold: null
+          };
+          var observer = new IntersectionObserver(function(entries) {

```

Comment:
I may be wrong but I think instead of one IntersectionObserver in which we handle the logic of determining the visible heading, we should create separate IOs for each heading so that whenever one becomes visible it triggers its callback function.

Reply:
I don't think that's necessary: https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver/observe but not an expert so will change if you really think this is needed. It does work as is in all browsers I've tested though (Chrome, Safari, Firefox, Edge, iOS Safari)